[GH-30460] BulkImport: User roles are not consistent with guest status

avasconcelos114 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->
This PR adds to the validation of guest accounts, by allowing users who are `system_guest` to be imported even when they have not been assigned to a team or channels. This fixes the current behavior where the import job fails and removes the need for workarounds listed in the issue.

I also took the liberty to:
- Rename variables, add comments, and adjust the flow so that it's more human readable and easier to maintain/understand (But DO let me know if anything should be reverted back or improved upon!)

#### Ticket Link
<!--
If applicable, please include both or either of the following links:

Fixes https://github.com/mattermost/mattermost/issues/XXX
Jira https://mattermost.atlassian.net/browse/MM-XXX
-->
Fixes https://github.com/mattermost/mattermost/issues/30460

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```release-note
NONE
```
-->
```release-note
Fixed issue where mmctl import jobs would fail when guest users were not members of any teams or channels
```

avasconcelos114 committed:
fix: Issue where a guest not being assigned to teams or channels would fail imports

- Added handling (as valid) for when guest is a system_guest but has no teams or channels
- Added handling (as valid) for when guest does has a team assigned but no channels
- Renamed counter vars for readability and edited flow to be based on a total count instead of per-team count

avasconcelos114 commented:
The behavioral changes can be summarized by the changes in the test cases, where we are now treating system guests as valid when they don't have team memberships (or they have them but haven't been added to any channels)

hanzei commented on server/channels/app/imports/import_validators.go:
Should `isValidGuestRoles` return a proper `error` instead of a boolean? At the moment, it is kind of hard to understand what is going wrong from an admin perspective.

avasconcelos114 replied:
My understanding of the flow is that during the import, `isValidGuestRoles` runs and always returns a boolean, which is then used to return an error in `ValidateUserImportData`

```go
	if !isValidGuestRoles(*data) {
		return model.NewAppError("BulkImport", "app.import.validate_user_import_data.guest_roles_conflict.error", nil, "", http.StatusBadRequest)
	}
```

hanzei reviewed (changes requested):
Thanks for the PR!

I would love to see an "E2E" test for the change:

- Create a guest
- Create an export
- Import on a fresh instance
- Confirm that the import completes and the guest is present

avasconcelos114 committed:
fix: Wrapping guest validation with isSystemGuest

- Fixing issue where a system_user would be mistakenly evaluated as a guest due to them not having teams assigned

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

isacikgoz reviewed (commented):
+1 on adding an e2e test for this one. Other than that, implementation looks okay to me.

avasconcelos114 commented:
Hi again @hanzei and @isacikgoz! Just wondering whether the approach for the E2E mentioned above makes sense or whether you had something else in mind

avasconcelos114 committed:
Merge branch 'master' into guest_import_validation

# Conflicts:
#	server/channels/app/imports/import_validators.go
#	server/channels/app/imports/import_validators_test.go

pavelzeman commented:
GIthub clean up - this has been opened for a while, are we still doing this?
Would it make sense to either close or move to Draft?

avasconcelos114 commented:
Closing this as it's gone stale and it would make more sense to just re-implement it and open a fresh PR
